src/import/kv_importer.rs

```diff
@@ -0,0 +1,337 @@
+    }
+
+    fn insert(&self, token: Token, engine: Arc<EngineFile>) {
+        let mut inner = self.inner.lock().unwrap();
+        assert!(inner.clients.insert(token, engine).is_none());
+    }
+
+    pub fn remove(&self, token: Token) -> Option<Arc<EngineFile>> {
+        let mut inner = self.inner.lock().unwrap();
+        inner.clients.remove(&token)
+    }
+
+    pub fn open(&self, uuid: Uuid) -> Result<()> {
+        let mut inner = self.inner.lock().unwrap();
+        if inner.engines.contains_key(&uuid) {
+            return Ok(());
+        }
+
+        match self.dir.open(uuid) {
+            Ok(engine) => {
+                info!("open {:?}", engine);
+                inner.engines.insert(uuid, Arc::new(engine));
+                Ok(())
+            }
+            Err(e) => {
+                error!("open {}: {:?}", uuid, e);
+                Err(e)
+            }
+        }
+    }
+
+    /// Attach to the engine.
+    /// Clients with the same uuid will share the same engine.
+    pub fn attach(&self, token: Token, uuid: Uuid) -> Result<()> {
+        let mut inner = self.inner.lock().unwrap();
+        if inner.clients.contains_key(&token) {
+            return Err(Error::TokenExists(token));
+        }
+
+        let engine = match inner.engines.get(&uuid) {
+            Some(engine) => Arc::clone(engine),
+            None => return Err(Error::EngineNotFound(uuid)),
+        };
+
+        info!("attach {:?}", engine);
+        inner.clients.insert(token, engine);
+        Ok(())
+    }
+
+    pub fn detach(&self, token: Token) -> Result<()> {
+        match self.remove(token) {
+            Some(engine) => {
+                info!("detach {:?}", engine);
+                Ok(())
+            }
+            None => Err(Error::TokenNotFound(token)),
+        }
+    }
+
+    pub fn write(&self, token: Token, batch: WriteBatch) -> Result<usize> {
+        match self.remove(token) {
+            Some(engine) => match engine.write(batch) {
+                Ok(v) => {
+                    self.insert(token, engine);
+                    Ok(v)
+                }
+                Err(e) => {
+                    error!("write {:?}: {:?}", engine, e);
+                    Err(e)
+                }
+            },
+            None => Err(Error::TokenNotFound(token)),
+        }
+    }
+
+    /// Close the engine.
+    /// Engine can not be closed when it is writing.
+    pub fn close(&self, uuid: Uuid) -> Result<()> {
+        let mut engine = {
+            let mut inner = self.inner.lock().unwrap();
+            let engine = match inner.engines.remove(&uuid) {
+                Some(engine) => engine,
+                None => return Err(Error::EngineNotFound(uuid)),
+            };
+            match Arc::try_unwrap(engine) {
+                Ok(engine) => engine,
+                Err(engine) => {
+                    inner.engines.insert(uuid, engine);
+                    return Err(Error::EngineInUse(uuid));
```

Comment:
The usage of engine is actually maintained by whether or not there are references to the engine. However semantically they are not exactly the same concept. Are there better solutions?

Reply:
How can you ensure that "when the client finishes the write, it releases the reference"? If this is a guarantee according to implementations, I think it is somewhat weak because of no strong restrictions.

Reply:
if we can ensure the client uses the unique engine with UUID, so here only the same client can close the engine. If the client closes the engine, the following write will be failed, I think it is acceptable. So I think here we can close the engine directly, no need to check in used any more.

Reply:
Not really. Different clients from different machines can write to the same engine concurrently, and any client can close the engine if it knows that all writes are done. Moreover, I must use the unsafe way to close the engine with an `Arc` if I don't check the reference.